components/core/tests/test-ffi_KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,529 @@
+        auto const dict_vars{encoded_ast.get_dict_vars()};
+        auto const encoded_vars{encoded_ast.get_encoded_vars()};
+        return Value{FourByteEncodedTextAst{logtype, dict_vars, encoded_vars}};
+    }
+    FAIL("Underlying value type not supported!");
+
+    // Control flow should never reach here. Adding a return to silent clang-tidy warnings.
+    return {};
+}
+
+auto clone_node_id_value_pairs(KeyValuePairLogEvent::NodeIdValuePairs const& node_id_key_value_pairs
+) -> KeyValuePairLogEvent::NodeIdValuePairs {
+    KeyValuePairLogEvent::NodeIdValuePairs result;
+    for (auto const& [node_id, optional_val] : node_id_key_value_pairs) {
+        if (optional_val.has_value()) {
+            result.emplace(node_id, clone_value(optional_val.value()));
+        } else {
+            result.emplace(node_id, std::nullopt);
+        }
+    }
+    return std::move(result);
+}
+}  // namespace
+
+TEST_CASE("ffi_Value_basic", "[ffi][Value]") {
+    Value const null_value;
+    test_value_is<std::monostate>(null_value);
+    test_value_get_immutable_view<std::monostate>(null_value, std::monostate{});
+
+    constexpr value_int_t cIntVal{1000};
+    Value const int_value{cIntVal};
+    test_value_is<value_int_t>(int_value);
+    test_value_get_immutable_view<value_int_t>(int_value, cIntVal);
+
+    constexpr value_float_t cFloatValue{1000.0001};
+    Value const float_value{cFloatValue};
+    test_value_is<value_float_t>(float_value);
+    test_value_get_immutable_view<value_float_t>(float_value, cFloatValue);
+
+    constexpr value_bool_t cBoolVal{false};
+    Value const bool_value{cBoolVal};
+    test_value_is<value_bool_t>(bool_value);
+    test_value_get_immutable_view<value_bool_t>(bool_value, cBoolVal);
+
+    constexpr std::string_view cStringVal{"This is a test string message"};
+    Value const string_value{string{cStringVal}};
+    test_value_is<string>(string_value);
+    test_value_get_immutable_view<string>(string_value, string{cStringVal});
+
+    Value const eight_byte_encoded_text_ast_value{
+            get_encoded_text_ast<eight_byte_encoded_variable_t>(cStringToEncode)
+    };
+    test_value_is<EightByteEncodedTextAst>(eight_byte_encoded_text_ast_value);
+    test_value_get_immutable_view<EightByteEncodedTextAst>(
+            eight_byte_encoded_text_ast_value,
+            get_encoded_text_ast<eight_byte_encoded_variable_t>(cStringToEncode)
+    );
+
+    Value const four_byte_encoded_text_ast_value{
+            get_encoded_text_ast<four_byte_encoded_variable_t>(cStringToEncode)
+    };
+    test_value_is<FourByteEncodedTextAst>(four_byte_encoded_text_ast_value);
+    test_value_get_immutable_view<FourByteEncodedTextAst>(
+            four_byte_encoded_text_ast_value,
+            get_encoded_text_ast<four_byte_encoded_variable_t>(cStringToEncode)
+    );
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+TEST_CASE("ffi_KeyValuePairLogEvent_create", "[ffi]") {
```

Comment:
You can reduce the amount of code for this test by using `SECTION` and even nested `SECTION` macros. Recall that CMake runs every `SECTION` from the beginning of the test, so the schema tree, node_id_value_pairs would be reset for each outer `SECTION`.

E.g., instead of cloning `valid_node_id_value_pairs`, you could just create it outside a `SECTION` and use it inside different `SECTION`s.

Reply:
1. But by using sections, you'd only run the things outside the section(s) from the beginning, right?
2. Testing the same instance of the schema tree across *all* of the tests seems a bit extra since it should be const in the class anyway (and if it wasn't const, you're still relying on finding a bug in the implementation with a set of test cases that isn't exhaustive). Imo, having less code in the test is more valuable than testing that corner case.

From my reading, you could do something like:

1. Create schema tree
2. Section (optional): Test empty id-value pairs
3. Section (optional): Test mismatched types
4. Section (optional): Test valid id-value pairs
   * Section: Test duplicate key conflicts
   * Section: Test implicit key conflicts
5. Section (optional): Test out of bound node ID

That would allow you to at least get rid of the `clone_node_id_value_pairs` function (and make the code a bit easier to read).